src/Features/Core/Portable/EmbeddedLanguages/Classification/AbstractEmbeddedLanguageClassificationService.cs

```diff
@@ -83,20 +80,26 @@ internal abstract class AbstractEmbeddedLanguageClassificationService :
                 _cancellationToken = cancellationToken;
             }
 
-            public void Recurse(SyntaxNode node)
+            public void VisitTokens(SyntaxNode node)
             {
-                _cancellationToken.ThrowIfCancellationRequested();
-                if (node.Span.IntersectsWith(_textSpan))
+                var stack = new Stack<SyntaxNodeOrToken>();
+                stack.Push(node);
+                while (!stack.IsEmpty())
                 {
-                    foreach (var child in node.ChildNodesAndTokens())
+                    _cancellationToken.ThrowIfCancellationRequested();
+                    var currentNodeOrToken = stack.Pop();
+                    if (currentNodeOrToken.Span.IntersectsWith(_textSpan))
                     {
-                        if (child.IsNode)
-                        {
-                            Recurse(child.AsNode()!);
-                        }
-                        else
+                        foreach (var child in currentNodeOrToken.ChildNodesAndTokens().Reverse())
```

Comment:
I just keep the same order as the original recursion because I don't have any context here.
If anyone tells me it's fine then this Reverse() call can be removed

Reply:
the order is important.  this ensures we produce the classifications in lexicographic order.

Reply:
Thanks for verifing this : )